packages/server/lib/util/socket_whitelist.ts

```diff
@@ -0,0 +1,46 @@
+import _ from 'lodash'
+import Debug from 'debug'
+import net from 'net'
+import { Request } from 'express'
+
+const debug = Debug('cypress:server:util:socket_whitelist')
+
+/**
+ * Utility to validate incoming, local socket connections against a list of
+ * expected client TCP ports.
+ */
+export class SocketWhitelist {
+  whitelistedLocalPorts: number[] = []
+
+  /**
+   * Add a socket to the whitelist.
+   */
+  add = (socket: net.Socket) => {
+    const { localPort } = socket
+
+    debug('whitelisting socket %o', { localPort })
+    this.whitelistedLocalPorts.push(localPort)
+
+    socket.once('close', () => {
+      debug('whitelisted socket closed, removing %o', { localPort })
+      this._remove(socket)
```

Comment:
i don't think this is ever tested - can we add an integration test?

Reply:
added a unit test, which highlighted that i needed to use `_.pull`, not `_.remove` :man_facepalming: